lib/internal/streams/compose.js

```diff
@@ -0,0 +1,297 @@
+'use strict';
+
+const pipeline = require('internal/streams/pipeline');
+const Duplex = require('internal/streams/duplex');
+const { createDeferredPromise } = require('internal/util');
+const { destroyer } = require('internal/streams/destroy');
+const from = require('internal/streams/from');
+const {
+  isStream,
+  isIterable,
+  isReadable,
+  isWritable
+} = require('internal/streams/utils');
+const {
+  PromiseResolve,
+} = primordials;
+const {
+  AbortError,
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_ARG_VALUE,
+    ERR_INVALID_RETURN_VALUE,
+    ERR_MISSING_ARGS,
+  },
+} = require('internal/errors');
+const assert = require('internal/assert');
+
+class ComposeDuplex extends Duplex {
+  constructor(options) {
+    super(options);
+
+    // https://github.com/nodejs/node/pull/34385
+
+    if (options?.readable === false) {
+      this._readableState.readable = false;
+      this._readableState.ended = true;
+      this._readableState.endEmitted = true;
+    }
+
+    if (options?.writable === false) {
+      this._writableState.writable = false;
+      this._writableState.ending = true;
+      this._writableState.ended = true;
+      this._writableState.finished = true;
+    }
+  }
+}
+
+module.exports = function compose(...streams) {
+  if (streams.length === 0) {
+    throw new ERR_MISSING_ARGS('streams');
+  }
+
+  for (let n = 0; n < streams.length; ++n) {
+    if (typeof streams[n] === 'function' && streams[n].length === 0) {
```

Comment:
Checking `streams[n].length` is imperfect here as it can miss weird but perfectly valid functions like:

```js
function myFunc(...args) {
  const source = args[0];
}

console.log(myFunc.length);   // 0
```

and...

```js
function myFunc(source = {}) {}

console.log(myFunc.length);   // 0
```

and...

```js
function myFunc() {
  const source = arguments[0];
}

console.log(myFunc.length);   // 0
```

Reply:
it's imperfect but I don' know how to do a better check, it's up to the user to pass the documented signature.

Reply:
Optionally we can just remove the check.